glean-core/tests/dual_labeled_counter.rs

```diff
@@ -0,0 +1,527 @@
+            ..Default::default()
+        },
+        Some(vec!["key1".into()]),
+        Some(vec!["category1".into()]),
+    );
+
+    let metric = dual_labeled_counter.get("key1", "category1");
+    metric.add_sync(&glean, 1);
+
+    let snapshot = StorageManager
+        .snapshot_as_json(glean.storage(), "store1", true)
+        .unwrap();
+
+    assert_eq!(
+        json!({
+            "dual_labeled_counter": {
+                "telemetry.dual_labeled_counter": {
+                    "key1": {
+                        "category1": 1
+                    }
+                }
+            }
+        }),
+        snapshot
+    );
+}
+
+#[test]
+fn can_use_multiple_labels() {
+    let (glean, _t) = new_glean(None);
+    let dual_labeled_counter = DualLabeledCounterMetric::new(
+        CommonMetricData {
+            name: "dual_labeled_counter".into(),
+            category: "telemetry".into(),
+            send_in_pings: vec!["store1".into()],
+            disabled: false,
+            lifetime: Lifetime::Ping,
+            ..Default::default()
+        },
+        None,
+        None,
+    );
+
+    let metric = dual_labeled_counter.get("key1", "category1");
+    metric.add_sync(&glean, 1);
+
+    let metric = dual_labeled_counter.get("key1", "category2");
+    metric.add_sync(&glean, 2);
+
+    let metric = dual_labeled_counter.get("key2", "category1");
+    metric.add_sync(&glean, 3);
+
+    let metric = dual_labeled_counter.get("key2", "category2");
+    metric.add_sync(&glean, 4);
+
+    let snapshot = StorageManager
+        .snapshot_as_json(glean.storage(), "store1", true)
+        .unwrap();
+
+    assert_eq!(
+        json!({
+            "dual_labeled_counter": {
+                "telemetry.dual_labeled_counter": {
+                    "key1": {
+                        "category1": 1,
+                        "category2": 2
+                    },
+                    "key2": {
+                        "category1": 3,
+                        "category2": 4
+                    }
+                }
+            }
+        }),
+        snapshot
+    );
+}
+
+#[test]
+fn can_record_error_for_submetric() {
+    let (glean, _t) = new_glean(None);
+    let dual_labeled_counter = DualLabeledCounterMetric::new(
+        CommonMetricData {
+            name: "dual_labeled_counter".into(),
+            category: "telemetry".into(),
+            send_in_pings: vec!["store1".into()],
+            disabled: false,
+            lifetime: Lifetime::Ping,
+            ..Default::default()
+        },
+        Some(vec!["key1".into()]),
+        Some(vec!["category1".into()]),
+    );
+
+    let metric = dual_labeled_counter.get("key1", "category1");
+    metric.add_sync(&glean, -1);
+
+    // Make sure that the errors have been recorded
+    assert_eq!(
+        Ok(1),
+        test_get_num_recorded_errors(&glean, metric.meta(), ErrorType::InvalidValue)
+    );
+}
+
+#[test]
+fn labels_are_checked_against_static_list() {
```

Comment:
Two test that are missing:

* Static keys, dynamic categories
* Dynamic keys, static categories